proposals/p6333.md

```diff
@@ -0,0 +1,586 @@
+
+The expectation is that C++ users will already be using a fully featured build
+system, such as CMake. Migration should be easier if users can retain their
+existing build system, particularly since a typical migration can be expected to
+mix both Carbon and C++ code.
+
+While Carbon could provide _both_ a separate compilation system _and_ a fully
+featured build system, a build system is a substantial undertaking and we expect
+C++ developers to already have one.
+
+### Don't support packaging directive to filename mappings
+
+Instead of making a mapping from packaging directives to filenames, we could
+generate a list specific to the `Core` package, and not expose that for other
+packages.
+
+We shouldn't manually maintain a mapping for the `Core` package; it should be
+automated. It's likely that whatever we do in this space, however we would
+support a mapping, would be of interest to small projects. It will probably be
+low cost for us to build support things other than `Core`, so we should just do
+that.
+
+### Distribute pre-compiled versions of Core files
+
+Instead of building object files for `Core` on demand, we could distribute them
+as part of Carbon. The upside of this is it would make builds a little faster;
+the downside is that we'd end up in more of a situation where supported target
+platforms were enumerated, or perhaps where special platforms could be built
+on-demand in a bespoke manner.
+
+We can probably add limited caching where it'd help, and support all platforms
+using similar logic that way with little performance penalty.
+
+### Require C++ files be passed to clang directly
+
+We could require C++ files to be passed to clang or carbon clang instead of
+carbon build, and similarly cc_library instead of carbon_library. Behind the
+scenes, clang will be equivalent to carbon clang, so Carbon will already be able
+to have control over how that compiles.
+
+For `carbon_library` in particular, we want to support situations where a `.cpp`
+file depends on a Carbon API file and the corresponding Carbon `impl` file
+depends on the corresponding `.h` file. This means a tight association that's
+well-served by having `carbon_library` support both.
+
+Similarly, we want `carbon build` to be able to take a combination of C++ and
+Carbon files, and decide what commands to run based on the file extension. Doing
+similar with `carbon compile` is then just an extension of the same desire to
+not make users choose which command to run based only on the filename.
+
+There may be some confusion because of `carbon_library` supporting both C++ and
+Carbon code. We are trying to think of this in terms of ecosystems, wherein the
+ecosystem includes both C++ and Carbon. But people already have a sense of what
+"C++" is in this context, so this proposal suggests using "Carbon" to be
+inclusive of C++ code. Hopefully then people can learn that the new thing works
+with both.
+
+What's really missing is a better name for this than "carbon".
```

Comment:
We could invent a name for the C, C++, and Carbon ecosystem, and then use that here. Just to throw out an idea, maybe the "CX ecosystem"? (as in, /C.*/) -- so `cx_library`, `cx_binary`.

Reply:
Per open discussion today, probably keeping with `carbon_*` and saying it's essentially the "carbon" ecosystem.